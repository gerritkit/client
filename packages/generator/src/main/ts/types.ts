export type TObjectType = {
  type: string
  wrapper?: string
}

export type TMethodInfo = {
  originalName: string
  methodName: string
  description: string
  method: string
  path: string
  inputs: {
    body?: TObjectType
    params?: string[][]
    args?: string[]
  }
  returnType: TObjectType
}

export type TSectionInfo = {
  titleSection: string
  methods: TMethodInfo[]
}

export type TType = Record<string, string[]>

export type TObjectTypeBucket = TObjectType[]
