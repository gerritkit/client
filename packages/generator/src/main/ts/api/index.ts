export { accessRightsEndpoints } from './access'
export {
  accountEndpoints,
  starEndpoints,
  defaultStarEndpoints,
} from './accounts'
export {
  attentionSetEndpoints,
  reviewerEndpoints,
  changeEndpoints,
  changeEditEndpoints,
  revisionEndpoints,
  revisionReviewerEndpoints,
} from './changes'
export { configEndpoints } from './config'
export { documentationSearchEndpoints } from './documentation'
export {
  groupEndpoints,
  groupMemberEndpoints,
  subgroupEndpoints,
} from './groups'
export { pluginEndpoints } from './plugins'
export {
  projectEndpoints,
  branchEndpoints,
  commitEndpoints,
  childProjectEndpoints,
  dashboardEndpoints,
  labelEndpoints,
  tagEndpoints,
} from './projects'
