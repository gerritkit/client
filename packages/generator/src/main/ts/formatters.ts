import { TMethodInfo } from './types'

export function addGlobalVariables(types: string[], code: string) {
  return `
  import axios from 'axios'
  import {
    ${types.join(', \n')}
     } from '../types/index'
  // NOTE: https://gerrit-review.googlesource.com/Documentation/rest-api.html#output
  const xssiPrefix = ')]}\\''
  const parseGerritResponse = (data: string) => JSON.parse(data.slice(xssiPrefix.length))
  ${code}
  `
}

export function generateType({
  typeName,
  fields,
}: {
  typeName: string
  fields: Array<{ name: string; required: boolean }>
}) {
  return `
    export type T${typeName} = {
      ${fields
        .map(
          ({ name, required }) =>
            `${name.replace('`', '')}${required ? '' : '?'}: any,`,
        )
        .join('\n')}
    };
  `
}

export function getPathArgsTypes(args?: string[]) {
  return args && args?.length > 0
    ? `args: {${args.map((el) => `${el}: string`).join(', ')}},`
    : ''
}

function getParams(params?: string[][]) {
  return params && params.length > 0
    ? `params: {${params.map(([name]) => name).join(', ')}},`
    : ''
}

export function getParamsType(params?: string[][]) {
  return params && params.length > 0
    ? `params: {${params.map(([name]) => `${name}?: string`).join(', ')}},`
    : ''
}

export function getData(bodyType?: string) {
  return bodyType ? 'data,' : ''
}

export function getDataType(bodyType?: string) {
  return bodyType ? `data: ${bodyType},` : ''
}

export function getPathArgs(args?: string[]) {
  return args && args.length > 0 ? `args: {${args.join(', ')}},` : ''
}

export function generateFunction({
  methodName,
  method,
  inputs,
  path = '',
  returnType,
}: TMethodInfo) {
  const { args, body, params } = inputs
  const data = getData(formatType(body))
  const dataType = getDataType(formatType(body))
  const pathArgs = getPathArgs(args)
  const pathArgsTypes = getPathArgsTypes(args)

  const paramsInput = getParams(params)
  const paramsInputTypes = getParamsType(params)

  const empty = !data && !pathArgs && !paramsInput
  const argsStr = empty ? '' : `{${data} ${pathArgs} ${paramsInput} } :`
  const argsStrType = empty
    ? ''
    : `{${dataType} ${pathArgsTypes} ${paramsInputTypes}}`

  const formattedParams = `${
    params && params.map(([name, filed]) => `${filed}: ${name}`).join(',\n')
  }`

  return `
  async ${methodName} ( ${argsStr} ${argsStrType} ) {
    return axios({
      method: '${method}',
      url: \`\${baseUrl}${path}\`,
      auth,
      params: {
        ${formattedParams}
      },
      ${body ? 'data,' : ''}
    }).then(({data}) => parseGerritResponse(data) as ${formatType(returnType)})
  },
`
}

export function generateSectionCode({
  titleSection,
  methods,
}: {
  titleSection: string
  methods: string[]
}) {
  if (methods.length === 0) {
    return ''
  }

  return `
    export function ${titleSection} ({ baseUrl, auth }: { baseUrl: string, auth: {
      username: string
      password: string
    } }) {
      return {
        ${titleSection}: {
          ${methods.join('\n')}
        }
      }
    }
    `
}

export function formatType(
  data: { wrapper?: string; type: string } | undefined,
) {
  if (!data) {
    return
  }

  const { type, wrapper } = data
  if (type === 'any') return 'any'
  if (wrapper === 'map') return `Record<string, T${type}>`
  if (wrapper === 'list') return `T${type}[]`
  return `T${type}`
}

export function generateDocs(
  data: { titleSection: string; methods: TMethodInfo[] }[],
  types: Record<string, string>,
) {
  return `
${data.map((data) => {
  return `
## ${data.titleSection}
${data.methods
  .map(
    ({
      methodName,
      description,
      returnType,
      inputs: { body, params, args },
    }) => {
      return `
### nativeClient.${data.titleSection}.${methodName}(${
        !args && !body && !params ? `` : 'input: TInput'
      })

${
  !args && !body && !params
    ? ''
    : `
#### Arguments:
\`\`\`typescript
type TInput = {
  ${args ? getPathArgsTypes(args || []) : ''}
  ${body ? types[getDataType(formatType(body))] : ''}
  ${params ? getParamsType(params) : ''}
}
\`\`\`
`
}


#### Returns:
\`\`\`typescript
${
  types[returnType?.type]
    ? `${types[returnType?.type]}`
    : 'type TReturnType = any'
}
\`\`\`
${description ? description.toString() : ''}
`
    },
  )
  .join('')}
`
})}
`
}
