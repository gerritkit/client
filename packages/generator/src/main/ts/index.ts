import axios from 'axios'
import cheerio from 'cheerio'
import { promises as fs } from 'fs'

import {
  addGlobalVariables,
  formatType,
  generateFunction,
  generateSectionCode,
  generateType,
  getDataType,
  getParamsType,
  getPathArgsTypes,
} from './formatters'
import { getSectionInfo, getSections, getTypes } from './parsers'
import { TMethodInfo } from './types'

const unsupportedSections = ['iDs', 'jSONEntities']

export async function getDocsApiVersion(url: string) {
  const html = await axios.get(url)
  const $ = cheerio.load(html.data)
  return $('#revnumber').first().text().split(' ')[1]
}

export async function generate(url: string) {
  const sections = await getSections(url)
  const types = sections
    .map((el) => getTypes(el))
    .filter((el) => el)
    .flat()

  const formattedTypes = types.map(generateType).join('\n')

  const typesForDocs = types.reduce((acc, el) => {
    acc[el.typeName] = generateType(el).slice('export type'.length)
    return acc
  }, {})

  const code = sections
    .map(getSectionInfo)
    .filter((section) => !unsupportedSections.includes(section.titleSection))

  const importTypes = [
    ...new Set([
      ...code.flatMap((el) => el.methods.map((q) => q.inputs.body?.type)),
      ...code.flatMap((el) => el.methods.map((q) => q.returnType?.type)),
    ]),
  ]
    .filter((el) => el && el !== 'any')
    .map((type) => `T${type}`)

  return {
    types: formattedTypes,
    docs: generateDocs(code, typesForDocs),
    code: addGlobalVariables(
      importTypes,
      code
        .map(({ titleSection, methods }) => ({
          titleSection,
          methods: methods.map(generateFunction),
        }))
        .map(generateSectionCode)
        .join('\n'),
    ),
  }
}

export async function generates(urls: [string, string][], path: string) {
  await fs.mkdir(`${path}/api/`)
  await fs.mkdir(`${path}/types/`)
  await fs.mkdir(`${path}/docs/`)
  await Promise.all(
    urls.map(async ([name, url]) => {
      const { code, types, docs } = await generate(url)
      await Promise.all([
        fs.writeFile(`${path}/api/${name}.ts`, code),
        fs.writeFile(`${path}/types/${name}.ts`, types),
        fs.writeFile(`${path}/docs/${name}.md`, docs),
      ])
    }),
  )
}

export function generateDocs(
  data: { titleSection: string; methods: TMethodInfo[] }[],
  types: Record<string, string>,
) {
  const a = data.map((data) => {
    return `
## ${data.titleSection}
${data.methods.map(
  ({ methodName, description, returnType, inputs: { body, params, args } }) => {
    return `
### native.${data.titleSection}.${methodName}(input: TInput)
#### inputs:
\`\`\`typescript
type TInput = {
  ${args ? getPathArgsTypes(args || []) : ''}
  ${body ? types[getDataType(formatType(body))] : ''}
  ${params ? getParamsType(params) : ''}
}
\`\`\`
#### returns:
\`\`\`typescript
${types[returnType?.type] ? `${types[returnType?.type]}` : ''}
\`\`\`
${description || ''}
`
  },
)}
`
  })
  return `
${a}
`
}
