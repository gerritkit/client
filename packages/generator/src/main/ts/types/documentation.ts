export type TDocResult = {
  title: any
  url: any
}
