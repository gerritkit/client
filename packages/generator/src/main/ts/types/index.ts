export {
  TAccessSectionInfo,
  TPermissionInfo,
  TPermissionRuleInfo,
  TProjectAccessInfo,
} from './access'
export {
  TAccountDetailInfo,
  TAccountExternalIdInfo,
  TAccountInfo,
  TAccountInput,
  TAccountNameInput,
  TAccountStatusInput,
  TAvatarInfo,
  TCapabilityInfo,
  TContributorAgreementInfo,
  TContributorAgreementInput,
  TDeletedDraftCommentInfo,
  TDeleteDraftCommentsInput,
  TDiffPreferencesInfo,
  TDiffPreferencesInput,
  TDisplayNameInput,
  TEditPreferencesInfo,
  TEmailInfo,
  TEmailInput,
  TGpgKeyInfo,
  TGpgKeysInput,
  THttpPasswordInput,
  TOAuthTokenInfo,
  TPreferencesInfo,
  TPreferencesInput,
  TProjectWatchInfo,
  TQueryLimitInfo,
  TSshKeyInfo,
  TStarsInput,
  TUsernameInput,
} from './accounts'
export {
  TCommitInfo,
  TAbandonInput,
  TActionInfo,
  TAddReviewerResult,
  TApprovalInfo,
  TAssigneeInput,
  TAttentionSetInfo,
  TAttentionSetInput,
  TBlameInfo,
  TChangeEditInput,
  TChangeEditMessageInput,
  TChangeInfo,
  TChangeInput,
  TChangeMessageInfo,
  TCherryPickInput,
  TCommentInfo,
  TCommentInput,
  TCommentRange,
  TCommitMessageInput,
  TContextLine,
  TDeleteChangeMessageInput,
  TDeleteCommentInput,
  TDeleteReviewerInput,
  TDeleteVoteInput,
  TDescriptionInput,
  TDiffContent,
  TDiffFileMetaInfo,
  TDiffInfo,
  TDiffIntralineInfo,
  TDiffWebLinkInfo,
  TEditFileInfo,
  TEditInfo,
  TFetchInfo,
  TFileInfo,
  TFixInput,
  TFixReplacementInfo,
  TFixSuggestionInfo,
  TGitPersonInfo,
  TGroupBaseInfo,
  THashtagsInput,
  TIncludedInInfo,
  TLabelInfo,
  TMergeableInfo,
  TMergeInput,
  TMergePatchSetInput,
  TMoveInput,
  TNotifyInfo,
  TPrivateInput,
  TProblemInfo,
  TPublishChangeEditInput,
  TPureRevertInfo,
  TPushCertificateInfo,
  TRangeInfo,
  TRebaseInput,
  TRelatedChangeAndCommitInfo,
  TRelatedChangesInfo,
  TRequirement,
  TRestoreInput,
  TRevertInput,
  TRevertSubmissionInfo,
  TReviewerInfo,
  TReviewerInput,
  TReviewerUpdateInfo,
  TReviewInfo,
  TReviewInput,
  TReviewResult,
  TRevisionInfo,
  TRobotCommentInfo,
  TRobotCommentInput,
  TRuleInput,
  TSubmitInfo,
  TSubmitInput,
  TSubmitRecord,
  TSubmittedTogetherInfo,
  TSuggestedReviewerInfo,
  TTopicInput,
  TTrackingIdInfo,
  TVotingRangeInfo,
  TWebLinkInfo,
  TWorkInProgressInput,
} from './changes'
export {
  TAccountsConfigInfo,
  TAuthInfo,
  TCacheInfo,
  TCacheOperationInput,
  TChangeConfigInfo,
  TChangeIndexConfigInfo,
  TCheckAccountExternalIdsInput,
  TCheckAccountExternalIdsResultInfo,
  TCheckAccountsInput,
  TCheckAccountsResultInfo,
  TCheckGroupsInput,
  TCheckGroupsResultInfo,
  TConfigUpdateEntryInfo,
  TConfigUpdateInfo,
  TConsistencyCheckInfo,
  TConsistencyCheckInput,
  TConsistencyProblemInfo,
  TDownloadInfo,
  TDownloadSchemeInfo,
  TEmailConfirmationInput,
  TEntriesInfo,
  TGerritInfo,
  THitRatioInfo,
  TIndexChangesInput,
  TIndexConfigInfo,
  TJvmSummaryInfo,
  TMemSummaryInfo,
  TPluginConfigInfo,
  TReceiveInfo,
  TServerInfo,
  TSshdInfo,
  TSuggestInfo,
  TSummaryInfo,
  TTaskInfo,
  TTaskSummaryInfo,
  TThreadSummaryInfo,
  TTopMenuEntryInfo,
  TTopMenuItemInfo,
  TUserConfigInfo,
} from './config'
export { TDocResult } from './documentation'
export {
  TGroupAuditEventInfo,
  TGroupInfo,
  TGroupInput,
  TGroupOptionsInfo,
  TGroupOptionsInput,
  TGroupsInput,
  TMembersInput,
} from './groups'
export { TPluginInfo, TPluginInput } from './plugins'
export {
  TAccessCheckInfo,
  TAutoCloseableChangesCheckInput,
  TAutoCloseableChangesCheckResult,
  TBanInput,
  TBanResultInfo,
  TBatchLabelInput,
  TBranchInfo,
  TBranchInput,
  TCheckProjectInput,
  TCheckProjectResultInfo,
  TCommentLinkInfo,
  TConfigInfo,
  TConfigInput,
  TConfigParameterInfo,
  TDashboardInfo,
  TDashboardInput,
  TDashboardSectionInfo,
  TDeleteBranchesInput,
  TDeleteLabelInput,
  TDeleteTagsInput,
  TGCInput,
  THeadInput,
  TIndexProjectInput,
  TInheritedBooleanInfo,
  TLabelDefinitionInfo,
  TLabelDefinitionInput,
  TLabelTypeInfo,
  TMaxObjectSizeLimitInfo,
  TProjectAccessInput,
  TProjectDescriptionInput,
  TProjectInfo,
  TProjectInput,
  TProjectParentInput,
  TReflogEntryInfo,
  TRepositoryStatisticsInfo,
  TSubmitTypeInfo,
  TTagInfo,
  TTagInput,
} from './projects'
