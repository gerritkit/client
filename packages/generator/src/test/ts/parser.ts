import {
  getSectionInfo,
  getSections,
  normaliseName,
  parseApiString,
} from '../../main/ts/parsers'
import { TSectionInfo } from '../../main/ts/types'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import stubDoc from './stub/api.json'

describe('parser', () => {
  const testDocsUrl = 'gerrit-host/Documentation/rest-api-projects.html'

  const mock = new MockAdapter(axios)
  mock.onGet(testDocsUrl).reply(200, stubDoc)
  it('getEndpoints ', async () => {
    const sections = await getSections(testDocsUrl)
    expect(sections.length).toBe(9)
  })

  it('getSectionInfo simple', async () => {
    const sections = await getSections(testDocsUrl)
    const section = sections[2]
    const res: TSectionInfo = getSectionInfo(section)
    expect(res).toMatchObject({
      titleSection: 'childProjectEndpoints',
      methods: [
        {
          originalName: 'List Child Projects',
          methodName: 'listChildProjects',
          inputs: { args: ['projectName'], params: [] },
          returnType: { type: 'ProjectInfo', wrapper: 'list' },
          description:
            '\nList the direct child projects of a project.\n\nAs result a list of ProjectInfo entries is\nreturned that describe the child projects.\n\nTo resolve the child projects of a project recursively the parameter\nrecursive can be set.\n\nChild projects that are not visible to the calling user are ignored and\nare not resolved further.\n',
          method: 'GET',
          path: '/projects/${projectName}/children/',
        },
        {
          originalName: 'Get Child Project',
          methodName: 'getChildProject',
          inputs: { args: ['projectName', 'projectName1'], params: [] },
          returnType: { type: 'ProjectInfo' },
          description:
            '\nRetrieves a child project. If a non-direct child project should be\nretrieved the parameter recursive must be set.\n\nAs response a ProjectInfo entity is returned that\ndescribes the child project.\n',
          method: 'GET',
          path: '/projects/${projectName}/children/${projectName1}',
        },
      ],
    })
  })

  it('getSectionInfo with params', async () => {
    const sections = await getSections(testDocsUrl)
    const section = sections[0]
    const res = getSectionInfo(section)
    expect(res).toMatchSnapshot()
  })

  it('normaliseName', async () => {
    expect(normaliseName('List Projects')).toBe('listProjects')
    expect(normaliseName('Project')).toBe('project')
    expect(normaliseName('foo, Bar .,. baz')).toBe('fooBarbaz')
  })

  it('parseApiString', async () => {
    expect(
      parseApiString(
        'GET /projects/{project-name}/children/{project-name}/children/{project-name}',
      ),
    ).toMatchObject({
      method: 'GET',
      args: ['projectName', 'projectName1', 'projectName2'],
      path:
        '/projects/${projectName}/children/${projectName1}/children/${projectName2}',
    })

    expect(
      parseApiString('GET /projects/{project-name}/branches/{branch-id}'),
    ).toMatchObject({
      method: 'GET',
      args: ['projectName', 'branchId'],
      path: '/projects/${projectName}/branches/${branchId}',
    })

    expect(
      parseApiString(
        'GET /projects/{project-name}/branches/{branch-id}/files/{file-id}/content',
      ),
    ).toMatchObject({
      method: 'GET',
      args: ['projectName', 'branchId', 'fileId'],
      path:
        '/projects/${projectName}/branches/${branchId}/files/${fileId}/content',
    })

    expect(
      parseApiString(
        'GET /projects/{project-name}/branches/{branch-id}/reflog',
      ),
    ).toMatchObject({
      method: 'GET',
      args: ['projectName', 'branchId'],
      path: '/projects/${projectName}/branches/${branchId}/reflog',
    })
  })
})
