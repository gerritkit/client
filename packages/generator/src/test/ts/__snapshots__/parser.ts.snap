// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser getSectionInfo with params 1`] = `
Object {
  "methods": Array [
    Object {
      "description": "
Lists the projects accessible by the caller. This is the same as
using the ls-projects command over SSH,
and accepts the same options as query parameters.

As result a map is returned that maps the project names to
ProjectInfo entries. The entries in the map are sorted
by project name.
",
      "inputs": Object {
        "args": Array [],
        "body": undefined,
        "params": Array [
          Array [
            "branch",
            "b",
          ],
          Array [
            "description",
            "d",
          ],
          Array [
            "limit",
            "n",
          ],
          Array [
            "prefix",
            "p",
          ],
          Array [
            "regex",
            "r",
          ],
          Array [
            "skip",
            "S",
          ],
          Array [
            "substring",
            "m",
          ],
          Array [
            "tree",
            "t",
          ],
          Array [
            "type",
            "type",
          ],
          Array [
            "state",
            "s",
          ],
        ],
      },
      "method": "GET",
      "methodName": "listProjects",
      "originalName": "List Projects",
      "path": "/projects/",
      "returnType": Object {
        "type": "ProjectInfo",
        "wrapper": "map",
      },
    },
    Object {
      "description": "
Queries projects visible to the caller. The
query string must be
provided by the query parameter. The start and limit parameters
can be used to skip/limit results.

As result a list of ProjectInfo entities is returned.
",
      "inputs": Object {
        "args": Array [],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "queryProjects",
      "originalName": "Query Projects",
      "path": "/projects/",
      "returnType": Object {
        "type": "ProjectInfo",
        "wrapper": "list",
      },
    },
    Object {
      "description": "
Retrieves a project.

As response a ProjectInfo entity is returned that
describes the project.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "getProject",
      "originalName": "Get Project",
      "path": "/projects/\${projectName}",
      "returnType": Object {
        "type": "ProjectInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Creates a new project.

In the request body additional data for the project can be provided as
ProjectInput.

As response the ProjectInfo entity is returned that
describes the created project.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "PUT",
      "methodName": "createProject",
      "originalName": "Create Project",
      "path": "/projects/\${projectName}",
      "returnType": Object {
        "type": "ProjectInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Retrieves the description of a project.

If the project does not have a description an empty string is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "getProjectDescription",
      "originalName": "Get Project Description",
      "path": "/projects/\${projectName}/description",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Sets the description of a project.

The new project description must be provided in the request body inside
a ProjectDescriptionInput entity.

As response the new project description is returned.

If the description was deleted the response is “204 No Content”.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "ProjectDescriptionInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "PUT",
      "methodName": "setProjectDescription",
      "originalName": "Set Project Description",
      "path": "/projects/\${projectName}/description",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Deletes the description of a project.

A commit message can be provided in the request body as a
ProjectDescriptionInput entity.
Historically, this method allowed a body in the DELETE, but that behavior is
deprecated.
In this case, use PUT instead.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "ProjectDescriptionInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "DELETE",
      "methodName": "deleteProjectDescription",
      "originalName": "Delete Project Description",
      "path": "/projects/\${projectName}/description",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Retrieves the name of a project’s parent project. For the
All-Projects root project an empty string is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "getProjectParent",
      "originalName": "Get Project Parent",
      "path": "/projects/\${projectName}/parent",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Sets the parent project for a project.

The new name of the parent project must be provided in the request body
inside a ProjectParentInput entity.

As response the new parent project name is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "ProjectParentInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "PUT",
      "methodName": "setProjectParent",
      "originalName": "Set Project Parent",
      "path": "/projects/\${projectName}/parent",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Retrieves for a project the name of the branch to which HEAD points.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "getHEAD",
      "originalName": "Get HEAD",
      "path": "/projects/\${projectName}/HEAD",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Sets HEAD for a project.

The new ref to which HEAD should point must be provided in the
request body inside a HeadInput entity.

As response the new ref to which HEAD points is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "HeadInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "PUT",
      "methodName": "setHEAD",
      "originalName": "Set HEAD",
      "path": "/projects/\${projectName}/HEAD",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Return statistics for the repository of a project.

The repository statistics are returned as a
RepositoryStatisticsInfo entity.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "getRepositoryStatistics",
      "originalName": "Get Repository Statistics",
      "path": "/projects/\${projectName}/statistics.git",
      "returnType": Object {
        "type": "RepositoryStatisticsInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Gets some configuration information about a project. Note that this
config info is not simply the contents of project.config; it generally
contains fields that may have been inherited from parent projects.

A ConfigInfo entity is returned that describes the
project configuration. Some fields are only visible to users that have
read access to refs/meta/config.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "getConfig",
      "originalName": "Get Config",
      "path": "/projects/\${projectName}/config",
      "returnType": Object {
        "type": "ConfigInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Sets the configuration of a project.

The new configuration must be provided in the request body as a
ConfigInput entity.

As response the new configuration is returned as a 
ConfigInfo entity.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "ConfigInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "PUT",
      "methodName": "setConfig",
      "originalName": "Set Config",
      "path": "/projects/\${projectName}/config",
      "returnType": Object {
        "type": "ConfigInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Run the Git garbage collection for the repository of a project.

Options for the Git garbage collection can be specified in the
request body as a GCInput entity.

The response is the streamed output of the garbage collection.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "GCInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "POST",
      "methodName": "runGC",
      "originalName": "Run GC",
      "path": "/projects/\${projectName}/gc",
      "returnType": Object {
        "type": "any",
      },
    },
    Object {
      "description": "
Marks commits as banned for the project. If a commit is banned Gerrit
rejects every push that includes this commit with
contains banned commit …​.

The commits to be banned must be specified in the request body as a
BanInput entity.

The caller must be project owner.

As response a BanResultInfo entity is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": Object {
          "type": "BanInput",
          "wrapper": undefined,
        },
        "params": Array [],
      },
      "method": "PUT",
      "methodName": "banCommit",
      "originalName": "Ban Commit",
      "path": "/projects/\${projectName}/ban",
      "returnType": Object {
        "type": "BanResultInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Lists the access rights for a single project.

As result a
ProjectAccessInfo
entity is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "GET",
      "methodName": "listAccessRightsforProject",
      "originalName": "List Access Rights for Project",
      "path": "/projects/\${projectName}/access",
      "returnType": Object {
        "type": "ProjectAccessInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
Sets access rights for the project using the diff schema provided by
ProjectAccessInput. Deductions are used to
remove access sections, permissions or permission rules. The backend will remove
the entity with the finest granularity in the request, meaning that if an
access section without permissions is posted, the access section will be
removed; if an access section with a permission but no permission rules is
posted, the permission will be removed; if an access section with a permission
and a permission rule is posted, the permission rule will be removed.

Additionally, access sections and permissions will be cleaned up after applying
the deductions by removing items that have no child elements.

After removals have been applied, additions will be applied.

As result a
ProjectAccessInfo
entity is returned.
",
      "inputs": Object {
        "args": Array [
          "projectName",
        ],
        "body": undefined,
        "params": Array [],
      },
      "method": "POST",
      "methodName": "addUpdateandDeleteAccessRightsforProject",
      "originalName": "Add, Update and Delete Access Rights for Project",
      "path": "/projects/\${projectName}/access",
      "returnType": Object {
        "type": "ProjectAccessInfo",
        "wrapper": undefined,
      },
    },
    Object {
      "description": "
This command runs access checks for other users. This requires the
View Access global
capability.

The result is a AccessCheckInfo entity
detailing the access of the given user for the given project,
project-ref, or project-permission-ref combination.
",
      "inputs": Object {
        "args": Array [],
        "body": undefined,
        "params": Array [
          Array [
            "account",
            "account",
          ],
          Array [
            "permission",
            "perm",
          ],
          Array [
            "ref",
            "ref",
          ],
        ],
      },
      "method": "GET",
      "methodName": "checkAccess",
      "originalName": "Check Access",
      "path": "/projects/MyProject/check.access",
      "returnType": Object {
        "type": "AccessCheckInfo",
        "wrapper": undefined,
      },
    },
  ],
  "titleSection": "projectEndpoints",
}
`;
